src/cfnlint/conditions.py

```diff
@@ -0,0 +1,316 @@
+"""
+  Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+  Permission is hereby granted, free of charge, to any person obtaining a copy of this
+  software and associated documentation files (the "Software"), to deal in the Software
+  without restriction, including without limitation the rights to use, copy, modify,
+  merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
+  permit persons to whom the Software is furnished to do so.
+
+  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+  INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+  PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+  HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+  SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+"""
+import hashlib
+from copy import copy
+import json
+import logging
+import six
+
+LOGGER = logging.getLogger(__name__)
+
+
+def get_hash(obj):
+    """ Return a hasl of an object """
+    return hashlib.sha1(json.dumps(obj, sort_keys=True).encode('utf-8')).hexdigest()
+
+
+class EqualsValue(object):
+    """ holds the values of a equals """
+    Function = None
+    String = None
+
+    def __init__(self, value):
+        if isinstance(value, dict):
+            if len(value) == 1:
+                # Save hashes of the dict for consistency and sorting
+                self.Function = get_hash(value)
+        elif isinstance(value, six.string_types):
+            self.String = value
+
+    def __eq__(self, other):
+        return other in [self.Function, self.String]
+
+
+class Equals(object):
+    """ Equals operator """
+    Left = None
+    Right = None
+
+    def __init__(self, equals):
+
+        if isinstance(equals, list):
+            if len(equals) == 2:
+                self.Left = EqualsValue(equals[0])
+                self.Right = EqualsValue(equals[1])
+
+    def test(self, scenarios):
+        """ Do an equals based on the provided scenario """
+        for scenario, value in scenarios.items():
+            if scenario == self.Left:
+                return value == self.Right
+            if scenario == self.Right:
+                return value == self.Left
+
+        return None
+
+
+class Condition(object):
+    """ Individual condition """
+    And = None
+    Or = None
+    Not = None
+    Equals = None
+    Influenced_Equals = None
```

Comment:
I don't know what this means. Can we explain the concept of `Influenced_Equals`?

Reply:
Basically all conditions eventually get down to equals.  So Influenced Equals includes all the Equals that make up a Condition.  So if there is sub condition from an and/or/not then we include its equals in the influenced equals.  That way can compare the base components of a condition... for instance all these are all related back to Ref myEnvironment either directly or indirectly.  
```
  isProduction: !Equals [!Ref 'myEnvironment', 'Prod']
  isNotProduction: !Not [{'Condition': 'isProduction'}]
  isProductionOrStaging: !Or [{'Condition': 'isProduction'}, {'Fn::Equals': [!Ref 'myEnvironment', 'Stage']}]
```